File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        product = []
        product.append(1)
        prev = nums[0]

        for i in range(1, len(nums)):
            product.append(prev)
            prev *= nums[i]

        prev = nums[len(nums)-1]
        for i in range(len(nums)-2, -1, -1):
            product[i] *= prev
            prev *= nums[i]

        return product

File: 238-product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] ans = new int[nums.length];
        int prevProd = 1;

        for(int i = 0; i < nums.length; i++) {
            ans[i] = prevProd;
            prevProd *= nums[i];
        }

        prevProd = 1;
        for(int i = nums.length-1; i >= 0; i--) {
            ans[i] *= prevProd;
            prevProd *= nums[i];
        }

        return ans;
    }
}